dbms/src/Common/LooseBoundedMPMCQueue.h

```diff
@@ -230,11 +241,22 @@ class LooseBoundedMPMCQueue
     }
 
     template <typename U>
-    ALWAYS_INLINE void pushFront(U && data)
+    ALWAYS_INLINE MPMCQueueResult pushFront(U && data)
     {
         Int64 memory_usage = get_auxiliary_memory_usage(data);
         queue.emplace_front(std::forward<U>(data), memory_usage);
         current_auxiliary_memory_usage += memory_usage;
+        if (push_callback)
+        {
+            if unlikely (!push_callback(queue.front().data))
```

Comment:
If `push_callback` returns false because `msg_channels_for_fine_grained_shuffle` was called `finish()`, it should not be possible to call `cancel` here?
Is it necessary for A to return the `MPMCQueueResult`?

Reply:
I think just return `MPMCQueueStatus::CANCELLED` and not change the queue status is ok? Because the caller should check the push result, and do something if the result is not `MPMCQueueStatus::OK`?